### PR TITLE
feat: vscode devcontainer

### PR DESCRIPTION
**Pull request by CheariX**
I added a [Remote Development Containers](https://code.visualstudio.com/docs/devcontainers/tutorial) in Visual Studio Code (VSCode).

Lots of people like to develop in Containers to have a clean system. With this PR, it is possible to work with al-folio without any installation (except for VS Code, its Remote Dev Container extension, and Docker).

Once you've opened the `al-folio` repository, a prompt will appear requesting to reopen the project within a container.

grafik

Upon doing so, Jekyll will automatically start within the container and prompt you to open the website's preview sidebar directly in VSCode or using your Browser. Additionally, it installs extensions for `liquid` and Prettier (`npx prettier`). Files are formatted using `al-folios`-prettier settings (`.prettierrc`) to streamline pull request submission.

Additionally, the performance seems to be much better compared to the `docker-compose`setup, see #2333.

### CODE REVIEW

**Comment by george-gca** [comment]
I don't think this is a good idea (to add the dev environment specifics to the repo) since some users like me don't use docker during development, and some not even VSCode. But I agree about adding your setup as a suggestion in the `INSTALL.md` file.

**Comment by CheariX** [comment]
The purpose of this proposal is to offer an alternative to the existing methods available. 

While there are several installation options provided in `INSTALL.md`, it might be beneficial to streamline and refine them, such as removing outdated or legacy ones. 

The VSCode setup is particularly interesting because it not only enables running Jekyll but also automatically integrates `prettier`. I have noticed that there have been instances where contributors forgot to apply prettier, and the documentation on how to use it remains unclear. Personally, it took me some time to get it functioning properly. Providing an example development container can serve as a starting point for users.

Remote Development has 5M+ installations,

> But I agree about adding your setup as a suggestion in the `INSTALL.md` file.

Do you mean to add the description to `INSTALL.md` without providing a `devcontainer.json`.
I would argue that this will not help anyone, because setting it up is not as easy as using it if a `devcontainer.json` is provided.

**Comment by saeub** [comment]
I don't think devcontainers are a feature specific to VS Code: https://www.jetbrains.com/help/idea/connect-to-devcontainer.html

**Comment by george-gca** [comment]
> Do you mean to add the description to INSTALL.md without providing a `devcontainer.json`.

Exactly. Maybe write the content of the file and how/where to create it in `INSTALL.md`. Like I said, some users like me don't use docker in their development environment, so this would add unnecessary files to our setup.

As for the install instructions we should remove the legacy information from the local install. It works, it is just not the best solution. But for some reasons a user might not be able to install docker in its environment, so it is an alternative.

**Comment by CheariX** [comment]
That's okay for me. It was just an offer to add another method to user. I'am not going to fight for dev container integration :-)

I changed the PR according to your input.

*** 

> Like I said, some users like me don't use docker in their development environment, so this would add unnecessary files to our setup.

In my humble opinion, I would argue that [over 23 million people use development containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) - it is the fourth most frequently installed Visual Studio Code extension. 

Regarding unnecessary files, it might be beneficial to eliminate one of those two Docker Compose files (for instance, `docker-compose-slim.yml`) and incorporate a dev container instead (this would maintain a consistent number of *additional/unnecessary* files while expanding the variety in how to utilize al-folio).

**Comment by george-gca** [comment]
In this matter I think it's best to include @alshedivat and @pourmand1376 on this discussion.

**Comment by pourmand1376** [comment]
Hi. 

I wish I've got involved in the conversation a little sooner. Since @george-gca has commented on this PR and my opinion is a little different. 

I like the idea of having dev containers. I installed it in `vscode` and used it and I am quite happy with it. 

But I am not totally happy with the way it is presented in `INSTALL.md`. We can just add the `.devcontainer/devcontainer.json` to our repository and whenever anyone opens up vscode, it would tell them to install the extension and guide them with the rest. 

And If the person doesn't know anything about containers, he can just ignore the prompt about installing devcontainers. He can also choose `Don't show this again`. 

So, I am pro with the idea of having support for devcontainers but I want not to pollute our `INSTALL.md` on how to do it.

**Comment by george-gca** [comment]
Do you know how to do this? This PR is still up to modifications @pourmand1376.

**Comment by CheariX** [comment]
So in summary, I see four possible solutions:

1. Close this PR without merging (in case there is no consent).
2. Add Devcontainer Instructions to `INSTALL.md` without providing `.devcontainer/devcontainer.json`.
    - This is the current state of this PR and matches @george-gca's idea. 
3. Add `.devcontainer/devcontainer.json` without mentioning/pulluting `INSTALL.md` 
    - This is what @pourmand1376 suggest, if I did not missunderstand him.
4. Add `.devcontainer/devcontainer.json` and a short description in `INSTALL.md`.
    - This was what we had initially with this PR

I hope that this correctly summarizes everything. We had one additional change (removing the legacy variant form `INSTALL.md`) but I think removing it aligns with what @pourmand1376 said ("but I want not to pollute our INSTALL.md").

I'm open for any solution. Tell me what you prefer.

**Comment by pourmand1376** [comment]
Thanks @george-gca, I know how to change it but I wanted us to first come to a conclusion about how to present this PR then we can change it. 

@CheariX 
Thanks for your clarifications. I think it may be best to explain devcontainers in one or two lines (very short) and then provide a link for people to understand it better. 

> Note: I am also a complete noob in this devcontainer thing. But When I opened up vscode with the `devcontainer.json` file in it, it prompted my to install the plugin and it did install everything necessary.

**Comment by george-gca** [comment]
Ok I can live with one single new unused file (at least for my setup) in the repo, no problem haha. So I am down for adding it in the repo.

Regarding the `INSTALL.md` part, 2 things:
1. I believe we can remove the `legacy` text, maybe change to not recommended or not supported? Idk, but I think we should keep it just in case someone can't install docker in its machine, or they simply prefers to use it locally.
2. I like the idea of not polluting it, but I think a simple explanation about this solution would be useful. Since you pointed out that it works not only for VSCode, we could add the modifications that @pourmand1376 commented, like the easy install part, but in `INSTALL.md` comment about this, maybe a couple of lines and a link to the dev containers site mentioning that it also supports other IDEs. Also since the docker solution is the recommended approach and dev containers align perfectly with it, we should move it up on the recommended setup.

> Regarding unnecessary files, it might be beneficial to eliminate one of those two Docker Compose files (for instance, docker-compose-slim.yml) and incorporate a dev container instead (this would maintain a consistent number of additional/unnecessary files while expanding the variety in how to utilize al-folio).

What about this part @pourmand1376? Any thoughts?

**Comment by pourmand1376** [comment]
Hi again, @george-gca 

I totally agree. 

I think we've arrived to a conclusion. Can you add the changes please, @CheariX?

**Comment by CheariX** [comment]
Okay, hopefully, I integerated everything according to the discussion.
Feel free do check and comment.

I decided to to a force push into this PR's changes because it had lots of back and forth. Now it is a clean single commit.

File: INSTALL.md [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
+- [Table of Contents](#table-of-contents)
 - [Installing and Deploying](#installing-and-deploying)
   - [Recommended Approach](#recommended-approach)
-  - [Local Setup on Windows](#local-setup-on-windows)
-  - [Local Setup using Docker (Recommended)](#local-setup-using-docker-recommended)
+  - [Local setup on Windows](#local-setup-on-windows)
+  - [Local setup using Docker (Recommended)](#local-setup-using-docker-recommended)
     - [Build your own docker image](#build-your-own-docker-image)
-  - [Local Setup (Legacy)](#local-setup-legacy)
+  - [Local Setup with VSCode Remote Containers](#local-setup-with-vscode-remote-containers)
```
Comment (george-gca) [inline INSTALL.md:9]: Please add again `Local Setup (Legacy, no longer supported)` in the toc

**Comment by george-gca** [comment]
> And If the person doesn't know anything about devcontainers, he can just ignore the prompt about installing devcontainers. He can also choose `Don't show this again` if he doesn't care about it anyway.

Is this part implemented?

**Review by george-gca — CHANGES_REQUESTED** [review]

File: INSTALL.md (hunk 1 above)
Comment (CheariX) [reply INSTALL.md]: Again a good catch. Also the text "with VSCode Remote Containers" was wrong.
Now updated automatically using the  devcontainer markdown extension

**Comment by CheariX** [comment]
> Is this part implemented?

The button is a VSCode feature, see the Screenshot in OP.

Would you like me to include this information in the `INSTALL.md` file, or perhaps you have a specific modification in mind?

File: INSTALL.md (hunk 1 above)
Comment (george-gca) [reply INSTALL.md]: I use GitHub copilot, so it does this part for me.

**Comment by george-gca** [comment]
No, it is fine. Just fixed a few capitalization changes. Thanks for your contribution.

**Comment by pourmand1376** [comment]
Thank you guys. @george-gca @CheariX

**Pull request merged**